xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,374 @@
+		closed:          grpcsync.NewEvent(),
+		pendingState: balancer.State{
+			ConnectivityState: connectivity.Connecting,
+			Picker:            base.NewErrPicker(balancer.ErrNoSubConnAvailable),
+		},
+	}
+}
+
+type gracefulSwitchBalancer struct {
+	bOpts balancer.BuildOptions
+	cc    balancer.ClientConn
+
+	mu               sync.Mutex
+	balancerCurrent  balancer.Balancer
+	balancerPending  balancer.Balancer
+	scToSubBalancer  map[balancer.SubConn]balancer.Balancer
+	pendingState     balancer.State
+	currentLbIsReady bool
+
+	swapMu            sync.Mutex
+	balRecentlyClosed balancer.Balancer
+
+	closed *grpcsync.Event
+}
+
+func (gsb *gracefulSwitchBalancer) updateState(bal balancer.Balancer, state balancer.State) {
+	if gsb.closed.HasFired() {
+		return
+	}
+
+	// I want this to not intersplice with other updateState() calls and for
+	// this to happen atomically (i.e. if this update isn't atomic writing to
+	// pending state concurrently and sending wrong one out for one
+	// updateState() call).
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+
+	gsb.swapMu.Lock()
+	gsb.balRecentlyClosed = nil
+	gsb.swapMu.Unlock()
+
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+
+	if bal == gsb.balancerPending {
+		// Cache the pending state and picker if you don't need to send it at
+		// this instant (i.e. the LB policy is not ready) you can send it later
+		// on an event like current LB exits READY.
+		gsb.pendingState = state
+		// "If the channel is currently in a state other than READY, the new
+		// policy will be swapped into place immediately." - Java
+		// implementation.
+		if state.ConnectivityState != connectivity.Connecting || !gsb.currentLbIsReady {
+			gsb.swap()
+		}
+	} else {
+		gsb.currentLbIsReady = state.ConnectivityState == connectivity.Ready
+		// In the specific case that the current lb exits ready, and there is a
+		// pending lb, you can forward the pending LB's cached State up to
+		// ClientConn and swap the pending into the current. "Otherwise, the
+		// channel will keep using the old policy until...the old policy exits
+		// READY." - Java
+		if !gsb.currentLbIsReady && gsb.balancerPending != nil {
+			gsb.swap()
+		} else {
+			// Even if there is a pending balancer waiting to be gracefully
+			// switched to, still forward current balancer updates to the Client
+			// Conn. Ignoring state + picker from the current would cause
+			// undefined behavior/cause the system to behave incorrectly from
+			// the current LB policies perspective. Also, the current LB is
+			// still being used by grpc to choose SubConns per RPC, and thus
+			// should use the most updated form of the current balancer.
+			gsb.cc.UpdateState(state)
+		}
+	}
+}
+
+// swap swaps out the current lb with the pending LB and updates the ClientConn.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) swap() {
+	gsb.cc.UpdateState(gsb.pendingState)
+	finished := make(chan struct{})
+	go func() {
```

Comment:
What's the purpose of this goroutine?

And if the parent goroutines waits for this to finish, it's the same as doing it inline.

Reply:
The purpose of this goroutine is to solve the race condition between Closing the current and Updating SubConnState on the current. (See my comment on your question as to why I only need swapMu synchronization with UpdateSubConnState()). Due to the possibility of calling back updateState() inline after UpdateSubConnState() on the current, this would induce a deadlock if we held the lock as we call down current.UpdateSubConnState(). Thus, I read into a local variable in UpdateSubConnState, unlocked the mutex, and then sent down the update. However, this causes a problem in UpdateSubConnState(), as now the balancer can be closed in between reading into local var pointer and updating it. Thus, I synchronized these two events (Close and UpdateSubConnState for child) with swapmu, which allows me to check if the balancer was closed before updating. Thus, I needed to not have a deadlock and needed to fork a goroutine. (This was Doug's suggestion to fix UpdateSubConnState() race by the way, which I agreed with).

Reply:
What I asked was, since `swap()` blocks and waits for the goroutine to finish, this is the same as without the goroutine.

```go
func a() {
  done := make(chan struct{})
  go func() { b(); done<-struct{}{} }()
  <-done
}
```

The above is the same as

```go
func a() {
  b()
}
```

Reply:
Got rid of blocking wait.